File: content/courses/parallel-computing-introduction/codes/mpi_io_example.f90
program mpiwrite
   use mpi_f08
   implicit none

   integer            :: i
   character(len=80)  :: arg
   integer            :: numargs

   integer            :: rank, nprocs
   integer            :: mpi_err
   integer, parameter :: root=0
   type(MPI_Status)   :: mpi_stat
   type(MPI_File)     :: fh
   integer            :: amode
   integer            :: tsize
   INTEGER(KIND=MPI_OFFSET_KIND) :: fsize, offset=0
   character(len=24)  :: fname
   character(len=1)   :: my_char
   integer            :: nreps

   character(len=:), allocatable :: gu,u
   character(len=256) :: rbuf

   ! check number of parameters and read in filename
   ! all ranks do this, avoids broadcast
   numargs=command_argument_count()
   if (numargs .lt. 1) then
      stop 'USAGE: output-file'
   else
      call get_command_argument(1,fname)
   endif

   !Initialize MPI
   call MPI_INIT()
   call MPI_COMM_SIZE(MPI_COMM_WORLD,nprocs)
   call MPI_COMM_RANK(MPI_COMM_WORLD,rank)

   my_char=char(iachar('A')+rank)

   amode=ior(MPI_MODE_CREATE, MPI_MODE_WRONLY)
   call MPI_FILE_OPEN(MPI_COMM_WORLD,trim(fname),amode,MPI_INFO_NULL,fh,mpi_err)

   if ( mpi_err /= MPI_SUCCESS) then
       call MPI_FINALIZE()
       stop "Unable to open MPI file for reading"
   endif

   nreps=20
   do i=1,nreps
      offset=rank+(i-1)*nprocs
      call MPI_FILE_WRITE_AT(fh, offset, my_char, 1, MPI_CHARACTER, mpi_stat)
   enddo

   call MPI_FILE_CLOSE(fh)

   ! Read as ordinary file
   ! Still can't read variable-length strings in Fortran
   if (rank==root) then
       rbuf=' '
       open(unit=10,file=fname,status='unknown')
       read(10,*) rbuf
       gu=trim(rbuf)
       write(*,'(a,a)') "Read at root ",gu
       close(10)
   endif

   !All processes read entire file
   !Blocking so will wait for root to finish above read
   amode=MPI_MODE_RDONLY
   call MPI_FILE_OPEN(MPI_COMM_WORLD,trim(fname),amode,MPI_INFO_NULL,fh,mpi_err)
   if ( mpi_err /= MPI_SUCCESS) then
       call MPI_FINALIZE()
       stop "Unable to open MPI file for reading"
   endif

   !Reading into "big enough" buffer means we don't need to get the size,
   !but this is the syntax.  Note type of fsize in declarations
   call MPI_FILE_GET_SIZE(fh,fsize)
   call MPI_TYPE_SIZE(MPI_CHARACTER,tsize)
   if (rank==root) then
       write(*,*) "File size is ",fsize," Type size is ",tsize," bytes"
   endif

   rbuf=' '
   call MPI_FILE_READ_ALL(fh, rbuf, len(rbuf), MPI_CHARACTER, mpi_stat)
   gu=trim(rbuf)
   write(*,'(a,i4.4,a,a)') 'Full file at rank ',rank,' ',gu

   !Read back the MPI file portion for reach rank
   rbuf=' '
   amode=MPI_MODE_RDONLY
   call MPI_FILE_OPEN(MPI_COMM_WORLD,trim(fname),amode,MPI_INFO_NULL,fh,mpi_err)
   if ( mpi_err /= MPI_SUCCESS) then
       call MPI_FINALIZE()
       stop "Unable to open MPI file for reading"
   else
       do i=1,nreps
          call MPI_FILE_READ_AT(fh, offset, rbuf(i:i), 1, MPI_CHARACTER, mpi_stat)
       enddo
   endif
   write(*,'(a,i4.4,a,a)') 'Rank ',rank,' ',trim(rbuf)

   call MPI_FILE_CLOSE(fh)

   call MPI_Finalize()

end program

File: content/courses/parallel-computing-introduction/codes/mpi_io_example.cxx
#include <cstring>
#include <cstdio>
#include <iostream>
#include <fstream>
#include <string>
#include <mpi.h>

using namespace std;

int main (int argc, char *argv[]) {

    // Declarations for MPI
    int rank, nprocs;
    int errcode;
    MPI_Status mpi_stat;
    MPI_Info info;
    MPI_Offset offset;
    MPI_File fh;
    int root=0, tag=0;
    int mpi_err;

    // Check number of parameters and read in filename
    if (argc < 2) {
       printf ("USAGE:  %s output-file\n", argv[0]);
    exit(1);
    }
    const char *fname=argv[1];

    //Initialize MPI
    MPI_Init(&argc, &argv);
    MPI_Comm_size(MPI_COMM_WORLD,&nprocs);
    MPI_Comm_rank(MPI_COMM_WORLD,&rank);

    int ord=int('A')+rank;
    char my_char=(char)ord;

    //Fun with pointers (sometimes fh is dereferenced, sometimes not)
    int amode=MPI_MODE_CREATE | MPI_MODE_WRONLY;
    mpi_err=MPI_File_open(MPI_COMM_WORLD,fname,amode,MPI_INFO_NULL,&fh);

    if (mpi_err != MPI_SUCCESS) {
        MPI_Finalize();
        exit(2);
    }

    int nreps=20;
    for (int i=0;i<nreps;i++) {
        offset=rank+i*nprocs;
        MPI_File_write_at(fh,offset,&my_char,1,MPI_CHAR,&mpi_stat);
    }
    MPI_File_close(&fh);

    // Read back in as ordinary file
    // We don't care about efficiency here
    if (rank==root) {
       ifstream fp(fname, ios::in);
       string sstr((istreambuf_iterator<char>(fp)), istreambuf_iterator<char>());
    cout<<sstr<<endl;
    fp.close();
    }

    //All processes read entire file
    //Blocking so will wait for root to finish above read
    amode=MPI_MODE_RDONLY;
    mpi_err=MPI_File_open(MPI_COMM_WORLD,fname,amode,MPI_INFO_NULL,&fh);
    if ( mpi_err != MPI_SUCCESS) {
       MPI_Finalize();
       cout<<"Unable to open MPI file for reading\n";
       exit(2);
    }
    MPI_Offset fsize;
    int tsize;
    MPI_File_get_size(fh,&fsize);
    MPI_Type_size(MPI_CHAR,&tsize);
    if (rank==root) {
        cout<<"File size is "<<fsize<<" Type size is "<<tsize<<" bytes\n";
    }
    int nchar=fsize/tsize;
    char fbuf[nchar];

    MPI_File_read_all(fh, fbuf, nchar, MPI_CHAR, &mpi_stat);
    MPI_File_close(&fh);
    cout<<"Rank "<<rank<<" ";
    for (int i=0; i<nchar; ++i) {
        cout<<fbuf[i];
    }
    cout<<endl;

    //Read back in as MPI file
    amode=MPI_MODE_RDONLY;
    mpi_err=MPI_File_open(MPI_COMM_WORLD,fname,amode,MPI_INFO_NULL,&fh);

    if (mpi_err != MPI_SUCCESS) {
        MPI_Finalize();
        cout<<"Unable to open MPI file for reading\n";
        exit(2);
    }

    char rbuf[nreps];
    offset=0;
    for (int i=0;i<nreps;i++) {
        offset=rank+i*nprocs;
        MPI_File_read_at(fh,offset,&rbuf[i],1,MPI_CHAR,&mpi_stat);
    }

    cout<<"Rank "<<rank<<" ";
    for (int i=0; i<nreps; ++i) {
        cout<<rbuf[i];
    }
    cout<<endl;

    MPI_File_close(&fh);

    MPI_Finalize();

}
